be/src/pipeline/exec/partition_sort_source_operator.cpp

```diff
@@ -0,0 +1,107 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "partition_sort_source_operator.h"
+
+#include "pipeline/exec/operator.h"
+
+namespace doris {
+class ExecNode;
+class RuntimeState;
+
+namespace pipeline {
+
+OperatorPtr PartitionSortSourceOperatorBuilder::build_operator() {
+    return std::make_shared<PartitionSortSourceOperator>(this, _node);
+}
+
+Status PartitionSortSourceLocalState::close(RuntimeState* state) {
+    if (_closed) {
+        return Status::OK();
+    }
+    _shared_state->_previous_row = nullptr;
+    _shared_state->_partition_sorts.clear();
+    return PipelineXLocalState<PartitionSortDependency>::close(state);
+}
+
+Status PartitionSortSourceOperatorX::get_block(RuntimeState* state, vectorized::Block* output_block,
+                                               SourceState& source_state) {
+    RETURN_IF_CANCELLED(state);
+    auto& local_state = state->get_local_state(id())->cast<PartitionSortSourceLocalState>();
+    output_block->clear_column_data();
+    {
+        std::lock_guard<std::mutex> lock(local_state._shared_state->_buffer_mutex);
+        if (local_state._shared_state->_blocks_buffer.empty() == false) {
+            local_state._shared_state->_blocks_buffer.front().swap(*output_block);
+            local_state._shared_state->_blocks_buffer.pop();
+            //if buffer have no data, block reading and wait for signal again
+            if (local_state._shared_state->_blocks_buffer.empty()) {
+                local_state._dependency->block_reading();
+            }
+            return Status::OK();
+        }
+    }
+
+    // this is set by sink node using: local_state._dependency->set_ready_for_read()
+    if (local_state._dependency->is_ready_for_read()) {
```

Comment:
`_dependency->is_ready_for_read()` is always true because we run this operator only if this condition is true